### PR TITLE
Allow case sensitive local file rename

### PR DESCRIPTION
**Pull request by stevealexr**
## Description
The FAT filesystem in android sd card is case-insensitive. This allows case-sensitive renaming by checking whether the changed name and original name is equivalent case-insensitively and doing a double rename. Only the code for local file renaming is changed.
Not sure what indent to use.

#### Issue tracker   
Fixes [#3587](https://github.com/TeamAmaze/AmazeFileManager/issues/3587)

#### Automatic tests
- [ ] Added test cases
  
#### Manual tests
- [x] Done  
  
- Device: Resizable
- OS: Api 33

#### Build tasks success  

Successfully running following tasks on local:
- [ ] `./gradlew assembledebug`
- [ ] `./gradlew spotlessCheck`
These tasks failed even with unmodified source code so probably local configuration issues

### CODE REVIEW

**Comment by VishalNehra** [comment]
Can you run `./gradlew checkspotless` and apply suggestions to fix the build failure please?

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:483]: Need a fallback mechanism in case the second rename fails, fallback should attempt to rename back to original file name.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:449]: This is too less random. Please randomize a bit more. Plus, suggested to put this at the beginning of the filename instead.

**Review by VishalNehra — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:449]: Putting it at the beginning of filename requires paths which is only available for API level 26 or higher. Doing it manually by string manipulation is risky. I have changed the temporary constant to a 32-characters string generated from `!@#$%a-zA-z0-9`.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:483]: I added a simple check that attempt to rename back to original file name, not sure if it is sufficient.

**Comment by stevealexr** [comment]
> Can you run `./gradlew spotlessCheck` and apply suggestions to fix the build failure please?

Managed to run spotlessCheck after applying this [fix](https://github.com/diffplug/spotless/issues/834) but assemble debug failure is caused by something else. I am using jdk 11.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:444]: Why is the case sensitive rename different?

Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:444]: This approach only cares about the value changed and not the type of filesystem. Although it does unnecessary double renaming in case-sensitive filesystem, it doesn't need to check the type of filesystem. 

The 'localRename' function (can't think of a better name) is extracted from the original code to avoid code duplication.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:481]: Thanks, changes seem fine. It would be a good practice to LOG.warn the inner else condition in case second rename fails.
Also, codacy does seem to complain can you fix the issue:

Check warning on line 673 in app/src/main/java/com/amaze/filemanager/filesystem/Operations.java

Codacy Production
/ Codacy Static Code Analysis
app/src/main/java/com/amaze/filemanager/filesystem/Operations.java#L673
These nested if statements could be combined

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:444]: > Why is the case sensitive rename different? Shouldn't there just be filesystems that support case sensitive renames and filesystems that do not? And the renames be the same function?

I think this should be okay as we know for sure most file systems that android supports doesn't support case sensitivity (exception ext4 in which case it needs to be kernel enabled). But I think it would make a marginal performance improvement if second part of the AND condition in this is used first.

**Comment by VishalNehra** [comment]
> > Can you run `./gradlew spotlessCheck` and apply suggestions to fix the build failure please?
> 
> Managed to run spotlessCheck after applying this [fix](https://github.com/diffplug/spotless/issues/834) but assemble debug failure is caused by something else. I am using jdk 11.

I think it's building just fine now remotely. For your local builds, I think you can point to your local jre instead. Make the following change in gradle.properties (if you're in mac).
org.gradle.java.home=/Applications/Android Studio.app/Contents/jre/Contents/Home

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:481]: I added a log message and combined those if statements.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (stevealexr) [inline app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:461]: The errorCallback.launchSAF needs to be lifted out of this function or else another errorCallback will be called.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java]: I added comment on the purpose of isCaseSensitiveRename variable and refactored double rename part. Hopefully it is clearer.

File: app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/filesystem/Operations.java:446]: This _only_ depends on the case sensitivity of the filesystem.

Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java]: I change it to a filesystem case-sensitivity check by checking whether both upper case and lower case input name for file exists return true.

Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java]: This new solution seem to be costly solution and calculating this each time is redundant. If a filesystem is case insensitive then it'll always be case insensitive, so you can persist the result of this in shared preferences and use that from next time.

Comment (stevealexr) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java]: There are multiple local filesystems like sdcard storage and internal storage. The types of filesystem need to be stored with the mount points. Rolling back to previous solution seems easier.

Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/filesystem/Operations.java]: Sure please revert back to your old solution. Also note that the new code will double rename even when FS is insensitive and new file and old file names aren't similar at all. 
oldFile.getSimpleName().equalsIgnoreCase(newFile.getSimpleName())
              && !oldFile.getSimpleName().equals(newFile.getSimpleName());

Please also note that I've increased the bounty price for this issue to 10USD you can claim once this PR is merged from the linked issue.

**Review by VishalNehra — DISMISSED** [review]

**Review by VishalNehra — APPROVED** [review]

**Comment by VishalNehra** [comment]
Please claim bounty [here](https://app.bountysource.com/issues/113047903-renaming-and-only-changing-case-fails)

**Comment by stevealexr** [comment]
> Please claim bounty [here](https://app.bountysource.com/issues/113047903-renaming-and-only-changing-case-fails) @stevealexr

Screenshot 2022-11-18 093833

I can't submit the claim. No claim button.

**Pull request merged**
